File: scripts/capture_checkpoint.py
#!/usr/bin/env python3
"""Deterministic Visual Evidence Capture Harness for Heliocide Viewer.

Captures exact 1600x900 screenshots across standard checkpoints:
  - A_NORMAL (t=8.0s, FIRST_PERSON)
  - C_SHARD_GOD (t=40.0s, FIRST_PERSON, modal=true)
  - D_COLLAPSE (t=65.0s, FIRST_PERSON)
  - E_BREACH (t=90.0s, FIRST_PERSON)
  - F_SIEGE_WALL (t=112.0s, FIRST_PERSON)
  - G_STATION_LOSS (t=130.0s, CINEMATIC)
  - H_REPLAY (t=138.0s, EXTERIOR_INSPECTION)
"""

import argparse
import os
import subprocess
import time
from pathlib import Path

CHROME_BIN = "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome"
USER_DATA_DIR = Path(".chrome-headless-profile").resolve()

CHECKPOINTS = {
    "A_NORMAL": {"time": 8.0, "camera": "FIRST_PERSON", "modal": "false"},
    "C_SHARD_GOD": {"time": 40.0, "camera": "FIRST_PERSON", "modal": "true"},
    "D_COLLAPSE": {"time": 65.0, "camera": "FIRST_PERSON", "modal": "false"},
    "E_BREACH": {"time": 90.0, "camera": "FIRST_PERSON", "modal": "false"},
    "F_SIEGE_WALL": {"time": 112.0, "camera": "FIRST_PERSON", "modal": "false"},
    "G_STATION_LOSS": {"time": 130.0, "camera": "CINEMATIC", "modal": "false"},
    "H_REPLAY": {"time": 138.0, "camera": "EXTERIOR_INSPECTION", "modal": "false"},
}

def capture_screenshot(checkpoint_name: str, output_path: str, port: int = 5173, custom_time: float = None, custom_camera: str = None, custom_modal: str = None):
    cfg = CHECKPOINTS.get(checkpoint_name, {"time": 8.0, "camera": "FIRST_PERSON", "modal": "false"})
    t = custom_time if custom_time is not None else cfg["time"]
    cam = custom_camera if custom_camera is not None else cfg["camera"]
    mod = custom_modal if custom_modal is not None else cfg["modal"]

    url = f"http://localhost:{port}/?time={t}&camera={cam}&modal={mod}&paused=true"
    out = Path(output_path).resolve()
    out.parent.mkdir(parents=True, exist_ok=True)
    USER_DATA_DIR.mkdir(parents=True, exist_ok=True)

    cmd = [
        CHROME_BIN,
        "--headless=new",
        "--disable-gpu",
        "--hide-scrollbars",
        "--no-first-run",
        "--no-default-browser-check",
        f"--user-data-dir={USER_DATA_DIR}",
        "--force-device-scale-factor=1",
        "--window-size=1600,900",
        f"--virtual-time-budget=2000",
        f"--screenshot={out}",
        url
    ]

    res = subprocess.run(cmd, capture_output=True, text=True)
    if out.exists() and out.stat().st_size > 0:
        print(f"Captured {checkpoint_name} -> {out} ({out.stat().st_size} bytes)")
        return True
    else:
        print(f"Failed to capture {checkpoint_name}: stderr={res.stderr} stdout={res.stdout}")
        return False

def main():
    parser = argparse.ArgumentParser(description="Capture visual checkpoints")
    parser.add_argument("--checkpoint", default="A_NORMAL", choices=list(CHECKPOINTS.keys()) + ["ALL"])
    parser.add_argument("--out", default="screenshot.png")
    parser.add_argument("--port", type=int, default=5173)
    parser.add_argument("--time", type=float, default=None)
    parser.add_argument("--camera", default=None)
    parser.add_argument("--modal", default=None)
    args = parser.parse_args()

    if args.checkpoint == "ALL":
        for cp in CHECKPOINTS:
            target_out = f"docs/visual-evolution/baselines/{cp}.png"
            capture_screenshot(cp, target_out, port=args.port)
    else:
        capture_screenshot(args.checkpoint, args.out, port=args.port, custom_time=args.time, custom_camera=args.camera, custom_modal=args.modal)

if __name__ == "__main__":
    main()

File: scripts/capture_checkpoint.js
#!/usr/bin/env node
import puppeteer from 'puppeteer-core';
import fs from 'fs';
import path from 'path';

const CHROME_PATH = '/Applications/Google Chrome.app/Contents/MacOS/Google Chrome';

const CHECKPOINTS = {
  A_NORMAL: { time: 8.0, camera: 'FIRST_PERSON', modal: 'false' },
  C_SHARD_GOD: { time: 40.0, camera: 'FIRST_PERSON', modal: 'true' },
  D_COLLAPSE: { time: 65.0, camera: 'FIRST_PERSON', modal: 'false' },
  E_BREACH: { time: 90.0, camera: 'FIRST_PERSON', modal: 'false' },
  F_SIEGE_WALL: { time: 112.0, camera: 'FIRST_PERSON', modal: 'false' },
  G_STATION_LOSS: { time: 130.0, camera: 'CINEMATIC', modal: 'false' },
  H_REPLAY: { time: 138.0, camera: 'EXTERIOR_INSPECTION', modal: 'false' },
};

async function capture(checkpointName, outputPath, options = {}) {
  const cfg = CHECKPOINTS[checkpointName] || { time: 8.0, camera: 'FIRST_PERSON', modal: 'false' };
  const time = options.time !== undefined ? options.time : cfg.time;
  const camera = options.camera || cfg.camera;
  const modal = options.modal !== undefined ? options.modal : cfg.modal;
  const port = options.port || 5173;

  const url = `http://localhost:${port}/?time=${time}&camera=${camera}&modal=${modal}&paused=true`;

  const dir = path.dirname(path.resolve(outputPath));
  if (!fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true });
  }

  const browser = await puppeteer.launch({
    executablePath: CHROME_PATH,
    headless: 'new',
    args: [
      '--no-sandbox',
      '--disable-setuid-sandbox',
      '--disable-dev-shm-usage',
      '--disable-gpu',
      '--use-gl=angle',
      '--use-angle=swiftshader',
      '--hide-scrollbars'
    ],
    defaultViewport: {
      width: 1600,
      height: 900,
      deviceScaleFactor: 1
    }
  });

  try {
    const page = await browser.newPage();
    await page.goto(url, { waitUntil: 'domcontentloaded', timeout: 15000 });
    // Wait for WebGL canvas and render settle
    await page.waitForSelector('canvas', { timeout: 10000 });
    await new Promise((r) => setTimeout(r, 1500));

    await page.screenshot({ path: outputPath, type: 'png' });
    console.log(`Successfully captured ${checkpointName} -> ${outputPath}`);
    return true;
  } catch (err) {
    console.error(`Error capturing ${checkpointName}:`, err);
    return false;
  } finally {
    await browser.close();
  }
}

async function main() {
  const args = process.argv.slice(2);
  let checkpoint = 'A_NORMAL';
  let out = 'screenshot.png';
  let port = 5173;

  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--checkpoint') checkpoint = args[++i];
    else if (args[i] === '--out') out = args[++i];
    else if (args[i] === '--port') port = parseInt(args[++i], 10);
  }

  if (checkpoint === 'ALL') {
    for (const cp of Object.keys(CHECKPOINTS)) {
      const targetOut = `docs/visual-evolution/baselines/${cp}.png`;
      await capture(cp, targetOut, { port });
    }
  } else {
    await capture(checkpoint, out, { port });
  }
  process.exit(0);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
